### PR TITLE
Do kprobe.multi on addresses instead of symbols

### LINKED ISSUE #284: Failing to load on bpf-next: calipso_skbuff_delattr: opening perf event: cannot assign requested address
```
Opening kprobe calipso_skbuff_delattr: creating perf_kprobe PMU (arch-specific fallback for "calipso_skbuff_delattr"): token calipso_skbuff_delattr: opening perf event: cannot assign requested address
```

From https://github.com/cilium/pwru/actions/runs/6719026297/job/18259892769?pr=283

### PR DESCRIPTION
**Pull request by Asphaltt**
Fix #284 
    
If we do kprobe.multi, kprobe on functions' addresses instead of symbols.

### CODE REVIEW

**Comment by Asphaltt** [comment]
By checking the CI log of bpf-next-main, we can see the unsupported kprobe functions are:

```
2023-12-17T11:30:32.2516205Z + grep 'Skip unsupported kprobe func' /tmp/pwru-advanced-ipv4.status
2023-12-17T11:30:32.2530356Z 2023/12/17 11:30:05 Skip unsupported kprobe func: e1000_xmit_frame
2023-12-17T11:30:32.2531623Z 2023/12/17 11:30:05 Skip unsupported kprobe func: calipso_skbuff_setattr
2023-12-17T11:30:32.2532906Z 2023/12/17 11:30:05 Skip unsupported kprobe func: calipso_skbuff_delattr
2023-12-17T11:30:32.2533677Z 2023/12/17 11:30:05 Skip unsupported kprobe func: add_grhead
2023-12-17T11:30:32.2534554Z 2023/12/17 11:30:05 Skip unsupported kprobe func: help
```

**Comment by brb** [comment]
Thanks for the PR!

I think the long term solution for both `kprobe` and `kprobe.multi` is to use func addrs instead of symbols. This would require extending the `kprobe` attach method in the `cilium/ebpf` library (should be trivial though). Meanwhile, it is already supported for the `kprobe.multi`.

As a temporary solution we can definitely accept this PR for the `kprobe` backend. However, I think for the `kprobe.multi` it might be too slow - the binary search tries to attach kprobes multiple times.

**Comment by Asphaltt** [comment]
> Meanwhile, it is already supported for the kprobe.multi.

Great. I'm on it.

File: main.go [hunk 1]
```diff
@@ -315,14 +295,25 @@ func main() {
 			default:
 			}
 
-			opts := link.KprobeMultiOptions{Symbols: funcsByPos[pos]}
+			addrs := make([]uintptr, 0, len(fns))
+			for _, fn := range fns {
+				if addr, ok := addr2name.Name2AddrMap[fn]; ok {
```
Comment (brb) [inline main.go:300]: As this is no longer 1-1 mapping, maybe we could change `Name2AddrMap` to store `map[string][]uintptr`?

Comment (Asphaltt) [reply main.go:300]: Thanks for pointing it out. I confirm it on 6.5.0-rc4+ kernel:

```bash
# ./pwru --output-meta --output-tuple icmp
2023/12/20 16:23:42 Attaching kprobes (via kprobe-multi)...
2023/12/20 16:23:42 Attaching kprobe-multi for pos 3 (116 functions / 116 addrs)
2023/12/20 16:23:42 Attaching kprobe-multi for pos 1 (953 functions / 956 addrs)
2023/12/20 16:23:42 Attaching kprobe-multi for pos 4 (12 functions / 12 addrs)
2023/12/20 16:23:42 Attaching kprobe-multi for pos 2 (350 functions / 350 addrs)
2023/12/20 16:23:42 Attaching kprobe-multi for pos 5 (31 functions / 31 addrs)
```

**Comment by brb** [comment]
@Asphaltt :wave: I'm catching up on reviews. Is this PR ready for the review?

**Comment by Asphaltt** [comment]
Yes, it's ready.

**Review by brb — APPROVED** [review]
Thanks! I will merge once I fix #323

**Pull request merged**